File: python/ProteinTranslation.py
from itertools import takewhile
from textwrap import wrap

catalogue = {
    'AUG': 'Methionine',
    'UUC': 'Phenylalanine',
    'UUU': 'Phenylalanine',
    'UUA': 'Leucine',
    'UUG': 'Leucine',
    'UCU': 'Serine',
    'UCC': 'Serine',
    'UCA': 'Serine',
    'UCG': 'Serine',
    'UAC': 'Tyrosine',
    'UAU': 'Tyrosine',
    'UGC': 'Cysteine',
    'UGU': 'Cysteine',
    'UGG': 'Tryptophan',
}

def is_not_stop(pattern):
    return pattern not in('UAG', 'UAA', 'UGA')
def proteins(strand):
    print(wrap(strand,3))
    return [protein
            for pattern in takewhile(is_not_stop, wrap(strand, 3))
            for protein in (catalogue.get(pattern, None),)
            if protein]

File: javascript/ProteinTranslation.js
const codons = {
    Methionine: ['AUG'],
    Phenylalanine: ['UUU', 'UUC'],
    Leucine: ['UUA', 'UUG'],
    Serine: ['UCU', 'UCC', 'UCA', 'UCG'],
    Tyrosine: ['UAU', 'UAC'],
    Cysteine: ['UGU', 'UGC'],
    Tryptophan: ['UGG'],
    STOP: ['UAA', 'UAG', 'UGA']
}

function find(codon) {
    for(const name in codons) {
        if(codons[name].includes(codon)) {
            return name
        }
    }
    return false
}

export function translate(RNA) {
    if(!RNA) return []

    const names = RNA.match(/.{1,3}/g).map(find),
        stop = names.findIndex(n => n === 'STOP'),
        final = stop === -1 ? names : names.slice(0, stop)

    if (final.some(n => !n)) {
        throw new Error('Invalid codon')
    }

    return final
}